features/mbedtls/targets/TARGET_Silicon_Labs/crypto_sha.c

```diff
@@ -0,0 +1,449 @@
+
+    /* Load the data block(s) */
+    for ( size_t i = 0; i < blocks; i++ ) {
+        if ((uint32_t)(&data[i*64]) & 0x3)
+        {
+            uint32_t temp[16];
+            memcpy(temp, &data[i*64], 64);
+            CORE_ENTER_CRITICAL();
+            CRYPTO_QDataWrite(&crypto->QDATA1BIG, temp);
+            CORE_EXIT_CRITICAL();
+        }
+        else
+        {
+            CORE_ENTER_CRITICAL();
+            CRYPTO_QDataWrite(&crypto->QDATA1BIG, (uint32_t*) &data[i*64]);
+            CORE_EXIT_CRITICAL();
+        }
+
+        /* Process the data block */
+        CRYPTO_EXECUTE_3(  crypto,
+                           CRYPTO_CMD_INSTR_SHA,
+                           CRYPTO_CMD_INSTR_MADD32,
+                           CRYPTO_CMD_INSTR_DDATA0TODDATA1 );
+    }
+    CORE_ENTER_CRITICAL();
+    CRYPTO_DDataRead(&crypto->DDATA0, state);
+    CORE_EXIT_CRITICAL();
+
+    crypto_management_release( crypto );
+}
+
+#if defined(MBEDTLS_SHA256_ALT) && defined(MBEDTLS_SHA256_C)
+void mbedtls_sha256_init( mbedtls_sha256_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_sha256_context ) );
+}
+
+void mbedtls_sha256_free( mbedtls_sha256_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_sha256_context ) );
+}
+
+void mbedtls_sha256_clone( mbedtls_sha256_context *dst,
+                           const mbedtls_sha256_context *src )
+{
+    if ( dst != NULL && src != NULL ) {
+        *dst = *src;
+    }
+}
+
+/*
+ * SHA-256 context setup
+ */
+void mbedtls_sha256_starts( mbedtls_sha256_context *ctx, int is224 )
+{
+    ctx->total[0] = 0;
+    ctx->total[1] = 0;
+
+    if ( is224 != 0 ) {
+        ctx->is224 = true;
+        memcpy(ctx->state, init_state_sha224, 32);
+    } else {
+        ctx->is224 = false;
+        memcpy(ctx->state, init_state_sha256, 32);
+    }
+}
+
+void mbedtls_sha256_process( mbedtls_sha256_context *ctx, const unsigned char data[64] )
+{
+    crypto_sha_update_state( ctx->state, data, 1, CRYPTO_SHA2 );
+}
+
+/*
+ * SHA-256 process buffer
+ */
+void mbedtls_sha256_update( mbedtls_sha256_context *ctx, const unsigned char *input,
+                    size_t ilen )
+{
+    size_t fill;
+    uint32_t left;
+
+    if( ilen == 0 ) {
+        return;
+    }
+
+    left = ctx->total[0] & 0x3F;
+    fill = 64 - left;
+
+    ctx->total[0] += (uint32_t) ilen;
+    ctx->total[0] &= 0xFFFFFFFF;
+
+    if( ctx->total[0] < (uint32_t) ilen ) {
+        ctx->total[1]++;
+    }
+
+    if( left && ilen >= fill ) {
+        memcpy( (void *) (ctx->buffer + left), input, fill );
+        mbedtls_sha256_process( ctx, ctx->buffer );
+        input += fill;
+        ilen  -= fill;
+        left = 0;
+    }
+
+    while( ilen >= 64 ) {
+        size_t blocks = ilen / 64;
+        crypto_sha_update_state( ctx->state, input, blocks, CRYPTO_SHA2 );
+        input += blocks * 64;
+        ilen  -= blocks * 64;
```

Comment:
have you tested a use case of Doing twice update of 65 bytes, and then finish?
I think I understand what you did.
Isn't it better to have only one counter, that will only tell the size of the "residue buffer"? Once this "counter" will reach 64, you will call `crypto_sha_update_state` , and set this residue buffer size to zero